Deprecates Ember.ArrayProxy

trek opened the pull request:
Includes a legacy flag for Ember Data to continue to use Ember.ArrayProxy internally.
- [x] Deprecate
- [ ] PR a shim to Ember Data that enables `_LEGACY_EMBER_DATA_SUPPORT` flag
- [ ] Add deprecation entry at http://emberjs.com/deprecations/v2.x/#toc_ember-runtimearray-proxy

Commit:
Deprecates Ember.ArrayProxy

mixonic commented on packages/ember-runtime/lib/system/array_proxy.js:
```diff
@@ -373,6 +373,17 @@ var ArrayProxy = EmberObject.extend(MutableArray, {
     this._setupContent();
     this._setupArrangedContent();
+
+    if (!this._LEGACY_EMBER_DATA_SUPPORT) {
```
Hm, this isn't better on `ENV` maybe? Ala `Ember.ENV._ENABLE_LEGACY_CONTROLLER_SUPPORT` ([usage](https://github.com/emberjs/ember.js/search?utf8=%E2%9C%93&q=_ENABLE_LEGACY_CONTROLLER_SUPPORT&type=Code))

rwjblue replied:
I believe this is correct.  Basically, we are only deprecating when not used via Ember Data.  This is on a per instance basis, not a global ENV basis.

mixonic replied:
Ah, gotcha. :+1:

mixonic commented:
Can you flag the tests with `[DEPRECATED]` at the end of the names? This is sweet.

rwjblue commented:
Can we also get an entry in emberjs/website's deprecation guide?

btecu commented:
Why is this being deprecated?
Is there an alternative?

rwjblue commented:
@btecu - Great question! That is exactly what the deprecation guide entry should point out.

cibernox commented:
I'm also curious about the reason. I use it in a couple places in conjunction with `PromiseProxyMixin`. Is it for performance reasons?

miguelcobain commented:
> Why is this being deprecated?

:scream:

trek commented:
@btecu @cibernox  can you tell us your use cases? It will help us write alternatives in the deprecation guide.

workmanw commented:
I can't speak for @btecu or @cibernox but my use case is I have a custom `QueryModelArray` that I've implemented in my application which is modeled after ember-data's `RecordArray`. My implementation allows for infinite paging and a couple of other domain specific helpers.

I don't have to use `Ember.ArrayProxy` per say, I'm just following the pattern laid out by ember-data. I saw it mentioned that "Includes a legacy flag for Ember Data". As long as I can use that in my app as well, I'm happy.

------

One other note, we DO use `Ember.ObjectProxy` for buffered proxy patterns. I hope that guy isn't slated for deprecation.

btecu commented:
I have a dynamic table `component` and in some cases I want to display different data than the received. So if I the value receive `yes` from the `api` I might want to display a check mark instead. In some cases and this depends on the specific `model` I'll display a specific icon or value depending on the value and column. This is specifically `UI` and it doesn't belong in the `api`.

So I use `ArrayProxy` for the entire model and then inside (`objectAtContent`) I use `ObjectProxy` for each row of the table that I need customized.

cibernox commented:
The use case I'm aware now, is using `ArrayProxy` in conjunction with `PromiseProxyMixin` like this:
```js
const PromiseArray = Ember.ArrayProxy.extend(Ember.PromiseProxyMixin);

export default Ember.Component.extend({
  results: computed('foo', function() {
    // ...
    return PromiseArray.create({ promise, content: previousResults });
  })
})
```

That basically gives me a collection-like object that I can iterate that contains some initial elements whose `isPending` property is `true` and once the promise resolves, the content of the array proxy is swapped by the resolved value of the promise (also `isPending` becomes `false`).

I'm not saying this shouldn't be deprecated, I already have rough idea in my head about how to stop using it, I just was curious about the reasons. Performance? I understand there must be some penalty, but I never measured it.

raytiley commented:
I'm doing basically the same pattern as @cibernox in several places in my app.

miguelcobain commented:
Same here.

alexdiliberto commented:
Just to note, I have converted some ArrayControllers over using this pattern, however, I believe `Ember.computed.sort()` is the more viable alternative.

```js
  foo: Ember.computed('model', function() {
    return Ember.ArrayProxy.extend(Ember.SortableMixin).create({
      sortProperties: ['bar'],
      sortAscending: false,
      content: this.get('model')
    });
  }
```

barneywilliams commented:
We use it for a few cases where we just want to be notified of an item being added/removed from a collection. The issue is that when a computed prop is dependent on `someCollection.[]` that you only get a the CP re-evaled. Figuring out which element was added/removed requires a cache implementation to diff against.

This is similar to the Collection [add](http://backbonejs.org/#Collection-add)/[remove](http://backbonejs.org/#Collection-remove) events in Backbone, which is one thing I missed when moving to Ember.

We have also use this type of functionality to insert and item into a sorted list at the right spot, rather than doing a full sort on a full set of data. We have experienced very significant perf gains on using this for large sets of data.

stefanpenner commented:
@barneywilliams do you mean an array? Or some arbitrary collection. If it's just an array, normal arrays already track such changes

toranb commented:
I'm also interested in why this is going away - one of my data store libraries is using this heavily like ember-data was for basic find/fetch/filtering of models in the store

```js
export default ArrayProxy.extend({
    store: null,
    type: null,
    content: computed(function () {
        return Ember.A(this.get("source"));
    })
});
```

barneywilliams commented:
@stefanpenner I was referring to collections in general.. but arrays are usually what I have dealt with. i know that you can get a notification on when an array changes in size with 'someArray.[]', but that doesn't  let you know what item(s) were added/removed... unless there is a trick I don't know about?

stefanpenner commented:
@barneywilliams although they are very low level, and must be used with care array observers allow for this.

stefanpenner commented:
@toranb makes a good point, until we can subclass arrays decorating them is likely one of the better options. This doesn't mean these proxies can't be extracted, but we should take this into account.

eviltrout commented:
We use Array Proxies in Discourse and we don't use Ember Data. We have our own lightweight data API on top of Ember and we use it for `ResultSet` objects like @toranb does. 

We also use it in a half dozen other places based on a quick grep. I'm also curious what the rationale for this deprecation is?

stefanpenner commented:
> I'm also curious what the rationale for this deprecation is?

I believe the concern is that it is a leaky abstraction, but maybe those concerns need to be fully enumerated. Maybe we can address those issues.

I wonder if a counter balance should be applied to the kill all proxies with fire. I believe stateful proxies, or proxies that leak their abstraction are the pain point. We should likely explore this in more detail.

xkb commented:
I've used `ArrayProxy` with `PromiseProxyMixin` to implement @toranb's "Use case 1" which I beleive what will async routable components provide in the future? 

Only way I've found to synchronously load a template and then update the array proxy when the actual promise resolves. Pretty much a `DS.PromiseArray`.

stefanpenner commented:
> which I beleive what will async routable components provide in the future?

nah, thats orthogonal

stefanpenner commented:
> Only way I've found to synchronously load a template and then update the array proxy when the actual promise resolves. Pretty much a DS.PromiseArray.

I would like to (before removing those constructs) make the underlying system correctly understand promises. So ultimately making them pointless before removing them.

pixelhandler commented:
Please please, do not deprecate ArrayProxy or ObjectProxy these are super useful for promise proxy objects. Seriously there is a lack of diversity among data solutions in the Ember ecosystem. These three objects are critical to developing data persistence solutions.

pixelhandler commented:
@trek please add a checkbox and issue for a "way forward" perhaps "[ ] Create Addon for PromiseProxyMixin, ArrayProxy and Object Proxy"

rwjblue commented:
@pixelhandler - `PromiseProxyMixin` and `ObjectProxy` are completely unrelated to this PR.

pixelhandler commented:
@rwjblue ok so how about "[ ] Create Addon for ArrayProxy"

pixelhandler commented:
@toranb Per your suggestion about sharing use cases…

I use the ArrayProxy (and ObjectProxy) with the PromiseProxyMixin as a utility for related resources in a data persistence library like so: https://github.com/pixelhandler/ember-jsonapi-resources/blob/master/addon/utils/related-proxy.js#L50-L73 and https://github.com/pixelhandler/ember-jsonapi-resources/blob/master/addon/utils/related-proxy.js#L90-L92 

I would be happy to use these same objects if they lived in an Ember CLI addon and no longer were included as primitives in Ember.js proper. I think they are valuable to anyone working data persistence.

I am curious what the proposed "way forward" is?

stefanpenner commented:
One thing worth noting, is all basic proxies like promise + object make removing `.get` basically not possible, or at least not a :footprints: :gun:

stefanpenner commented:
> I am curious what the proposed "way forward" is?

Well, for example: One could decorate an array with additional functionality. making it one with the proxy, rather then indirectly a proxy. This is more or less one path nicely aligned with the platform, especially once we are able to subclass native arrays cross platform. Unfortunately until real subclassing is possible that approach isn't really that best. :(

opsb commented:
We're using it to implement [live queries](https://github.com/orbitjs/ember-orbit/blob/rethink_integration/addon/live-query.js) and hasMany relationships in [orbit.js](https://github.com/orbitjs/orbit.js). 

It's not something that I often reach for but when needed it is very useful. I'm not sure what the reasoning behind it's removal but I do see a couple of issues that could perhaps be addressed.

1) `content` should really be private
2) It assumes the ArrayProxy will be mutable, in the case of a result set it should be readonly (we worked around this with [ReadOnlyArrayProxy](https://github.com/orbitjs/ember-orbit/blob/rethink_integration/addon/system/read-only-array-proxy.js)).

martndemus commented:
I use the `ArrayProxy` in combination with `unknownProperty` & `setUnknownProperty` to make an array of `DS.Model` instances act as a single object while maintaining being an array of things.

Example: setting the 'admin' property on the array proxy will set the 'admin' property on all items in the array proxy.

fivetanley commented:
ArrayProxy is something we'd (or at least I'd) like to move away from in Ember Data. Or at least our current usage of it. I don't have an alternate implementation, but an ArrayProxy that had a hidden `content` (even if just `_content`), didn't support `willChangeContent/didChangeContent` or `arrangedContent` would make me pretty happy as the implementations there give us many bug reports.

I don't know if we'd be able to drop it because of backwards compat, but removing the most troubling aspects of ArrayProxy would help a lot.

trek commented:
All: This was great! You might have seen in the [notes on the last face to face](http://emberjs.com/blog/2016/01/23/core-team-face-to-face-january-2016.html#toc_ember-3-0-deprecations) that the core team is trying to trickle out deprecations on a more, *ahem*, tolerable time table during this major revision.

Part of that process means doing PRs for deprecations early so we can get community comments. I think the core team were occasionally surprised at some of the creative uses of public 1.x API and want to do a better job ensuring alternate APIs exist for all use cases this time around.

`ArrayProxy` likely won't be in 3.0, but we won't merge this PR until we've addressed the concerns here.

@btecu, et al
> Why is this being deprecated?

`ArrayProxy` (and proxies in general) make more sense in the 1.x/2.x programming model. Glimmer and other framework changes are going make them mostly unnecessary in 3.x

The three main uses for `ArrayProxy` that have come up are

1. Extra features and observable aware methods on Array-like objects

  As @stefanpenner mentions, we already extend regular Array objects to track mutations with methods like `pushObject`. The alternative is to use Array.


1. Maintaining sort order/filtering/etc

  For behaviors like maintaining sort order you can use `Ember.computed.sort()`, `Array#sortBy` or replace the existing array with a new array that contains the items you want, and sort yourself:

  ```
  actions: {
    addAnItem(item) {
      let array = this.get('items').slice(); // <- creating a new array, VERY IMPORTANT.
      array.push(item);
      array.sort(someCompareFunction);
      this.set('items', array);
    }
  }
  ```

1. Presenting a stable reference for `model()` resolution as part of a data layer.

  As @toranb et al mention having a stable ref allows for immediate resolution of `model` hooks. Although not _all_ specifically enumerated, the underlying problems this approach address:
    
    * Manual control of loading states. I'm guessing that this desire comes from wanting to paint *some* the view for a route, just not portions of the view that rely on the contents of the collection being loaded.

    * Allows model hooks to be non-blocking, so child hooks that don't need access to an ancestor's model can trigger immediately.

  We still need to decide on patterns for handling the above two. The solutions will depend on the specifics on additions coming future 2.x releases (like engines, routable components).

  The solution might still be something Proxy-like, but that solution will drop methods unrelated to being a stable ref. Namely: `content` would be private, no `arrangedContent`, no `willChangeContent`/`didChangeContent`, filtering and sorting would be resposibilities of your data layer or view layer.

------
Instead of deprecating `ArrayProxy`, we *could* deprecate all the features that are better handled elsewhere, but that guts `ArrayProxy` so much that it effectively becomes a new class.

To aid migration, would having multiple deprecation warnings per ArrayProxy be helpful? That is, would it be useful to know where you were using each of the use cases above?

opsb commented:
> The solution might still be something Proxy-like, but that solution will drop methods unrelated to being a stable ref. Namely: content would be private, no arrangedContent, no willChangeContent/didChangeContent, filtering and sorting would be resposibilities of your data layer or view layer.

This solution will solve it, but I thought I'd reiterate the scenario that I'm considering:

In orbit.js we stream updates to live queries(queries whose results are updated in realtime) through an [Observable](https://github.com/zenparsing/es-observable) (we're using RxJS under the hood). Ember-orbit uses an ArrayProxy to provide a reference to the latest array of results received from this Observable.

runspired commented:
I use a few custom array proxies as well, but I do often reach for this. It's an incredibly useful tool.  I think it would make sense at least to provide this as an addon.

raytiley commented:
>```js
> const PromiseArray = Ember.ArrayProxy.extend(Ember.PromiseProxyMixin);
> 
> export default Ember.Component.extend({
>   results: computed('foo', function() {
>     // ...
>     return PromiseArray.create({ promise, content: previousResults });
>   })
> })
>```

@trek forgive me if I missed something but I don't specifically see a way forward in your reply for this use case. The major place we are still using `ArrayProxy` in our app is to return array like objects from a computed property that can be used in `{{each}}`. 

This seems to always be a thing that comes up but there isn't a best practice that I am aware of. Maybe it would make more sense to just return a promise from the CP that will resolve into an array, and make `{{each}}`  yield to the inverse if it is passed an in flight promise object?

homu commented:
:umbrella: The latest upstream changes (presumably #12914) made this pull request unmergeable. Please resolve the merge conflicts.

trek commented:
@raytiley yeah, we generally need better handling for async flow control at the template layer. I'm not sure who suggested it (@stefanpenner maybe?), but an `{{await someProperty}}` helper might fill the same use you need here?

stefanpenner commented:
Ya, I believe many of the cases described here can be replaced with tools that are better suited for the task. Although those may not all be ready yet, hopefully soon they will be.

jrjohnson commented:
We've been using ember-data's DS.PromiseArray to do the work @raytiley is talking about in https://github.com/emberjs/ember.js/pull/12843#issuecomment-178279807 

So I wasn't too worried about this given the caveat that data would be excepted.  However it appears that PromiseArray is actually private in data.  Deprecating this doesn't bother me, but before that happens there should be a shared solution to the common problem of needing to put together a set of one or more promises into an array to get passed into a template.  Right now it is dirt simple to simple return a DS.PromiseArray from a CP which {{each}} will use and which can also be tested for isFulfilled to provide for a nice loading indication or pre-loaded state.

Maybe it is making DS.PromiseArray public https://github.com/emberjs/data/issues/4163 ?

pixelhandler commented:
Is this PR inactive? What is the path forward? To deprecate or not to deprecate?

Serabe commented:
Closing this PR as inactive. Thank you!
